File: src/main/java/com/tabnine/inline/listeners/InlineKeyListener.java
package com.tabnine.inline.listeners;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.Disposer;
import com.tabnine.inline.CompletionPreview;

import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

public class InlineKeyListener extends KeyAdapter {
    private final Editor editor;

    public InlineKeyListener(Editor editor) {
        this.editor = editor;
    }

    @Override
    public void keyReleased(KeyEvent event) {
        CompletionPreview preview = CompletionPreview.findCompletionPreview(editor);
        if (preview == null || preview.isCurrentlyNotDisplayingInlays()) return;

        int key = event.getKeyCode();
        // do not interfere with inline shortcuts
        if (key == KeyEvent.VK_ALT
                || key == KeyEvent.VK_OPEN_BRACKET
                || key == KeyEvent.VK_CLOSE_BRACKET
                || key == KeyEvent.VK_TAB
        )
            return;

        try {
            Disposer.dispose(preview);
        } catch (Throwable err) {
            Logger.getInstance(getClass()).warn("Error in Tabnine preview KeyListener", err);
        }
    }
}

File: src/main/java/com/tabnine/inline/listeners/InlineKeyListener.kt
package com.tabnine.inline.listeners

import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.Disposer
import com.tabnine.inline.CompletionPreview
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent

private val INLINE_SHORTCUTS =
    setOf(KeyEvent.VK_ALT, KeyEvent.VK_OPEN_BRACKET, KeyEvent.VK_CLOSE_BRACKET, KeyEvent.VK_TAB)

class InlineKeyListener(
    private val completionPreview: CompletionPreview
) : KeyAdapter(), Disposable {
    init {
        completionPreview.editor.contentComponent.addKeyListener(this)
        Disposer.register(completionPreview, this)
    }

    override fun keyReleased(event: KeyEvent) {
        if (INLINE_SHORTCUTS.contains(event.keyCode)) {
            return
        }

        Logger.getInstance(javaClass).warn("BOAZ: KeyListenere disposing completion. Event: " + event.toString())
        Disposer.dispose(completionPreview)
    }

    override fun dispose() {
        Logger.getInstance(javaClass).warn("BOAZ: KeyListenere disposed")
        completionPreview.editor.contentComponent.removeKeyListener(this)
    }
}
